src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -397,9 +456,31 @@ private Task CloseAsync(ConnectionContext connection)
         // if we're disposing.
         private async Task StopAsyncCore(bool disposing)
         {
-            // Block a Start from happening until we've finished capturing the connection state.
+            // StartAsync acquires the connection lock for the duration of the handshake.
+            // ReconnectAsync also acquires the connection lock for reconnect attempts and handshakes.
+            // Cancel the StopCts without acquiring the lock so we can short-circuit it.
+            _state.StopCts.Cancel();
```

Comment:
It does seem scary to cancel the StopCts outside the lock, but we don't need to worry about the connection ending up in a no-longer-startable state because [this logic](https://github.com/aspnet/AspNetCore/pull/9535/files#diff-f10b84c92f9edf5f451fd3213c4a6a3bR265) in StartAsync() will reset the StopCts if it fails in this way.